sdk/iot/hub/src/az_iot_hub_client_sas.c

```diff
@@ -14,110 +14,165 @@
 
-  *out_document = az_span_slice(document, 0, _az_span_diff(remainder, document));
+  *out_signature = az_span_slice(signature, 0, _az_span_diff(remainder, signature));
 
   return AZ_OK;
 }
 
-az_result az_iot_sas_token_generate(
-    az_span iothub_fqdn,
-    az_span device_id,
-    az_span signature,
-    int32_t expiry_time_secs,
+AZ_NODISCARD az_result az_iot_hub_client_sas_get_password(
+    az_iot_hub_client const* client,
+    az_span base64_hmac_sha256_signature,
+    uint32_t token_expiration_epoch_time,
     az_span key_name,
-    az_span sas_token,
-    az_span* out_sas_token)
+    char* mqtt_password,
+    size_t mqtt_password_size,
+    size_t* out_mqtt_password_length)
 {
-  AZ_PRECONDITION_VALID_SPAN(iothub_fqdn, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(signature, 1, false);
-  AZ_PRECONDITION(expiry_time_secs > 0);
-  AZ_PRECONDITION_VALID_SPAN(sas_token, 0, false);
-  AZ_PRECONDITION_VALID_SPAN(sas_token, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_sas_token);
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(base64_hmac_sha256_signature, 1, false);
+  AZ_PRECONDITION(token_expiration_epoch_time > 0);
+  AZ_PRECONDITION_NOT_NULL(mqtt_password);
+  AZ_PRECONDITION(mqtt_password_size > 0);
 
   // Concatenates: "SharedAccessSignature sr=" scope "&sig=" sig  "&se=" expiration_time_secs
   //               plus, if key_name size > 0, "&skn=" key_name
 
-  az_span sr_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SR);
-  az_span devices_string = AZ_SPAN_FROM_STR(SCOPE_DEVICES_STRING);
-  az_span sig_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SIG);
-  az_span se_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SE);
-  int32_t required_length = az_span_size(sr_string) + az_span_size(devices_string)
-      + az_span_size(sig_string) + az_span_size(se_string) + az_span_size(iothub_fqdn)
-      + az_span_size(device_id) + az_span_size(signature) + 5;
+  // This does not account for the size of `token_expiration_epoch_time`, which will be handled by az_span_u32toa.
+  int32_t required_size = 
+      az_span_size(sr_string) 
+      + (int32_t)sizeof(EQUAL_SIGN)
+      + az_span_size(client->_internal.iot_hub_hostname)
+      + az_span_size(devices_string)
+      + az_span_size(client->_internal.device_id)
+      + (int32_t)sizeof(AMPERSAND)
+      + az_span_size(sig_string) 
+      + (int32_t)sizeof(EQUAL_SIGN)
+      + az_span_size(base64_hmac_sha256_signature)
+      + (int32_t)sizeof(AMPERSAND)
+      + az_span_size(se_string) 
+      + (int32_t)sizeof(EQUAL_SIGN)
+      + u32toa_size(token_expiration_epoch_time)
+      + (int32_t)sizeof(STRING_NULL_TERMINATOR);
+
+  if (az_span_size(client->_internal.options.module_id) > 0)
+  {
+    required_size +=
+        az_span_size(modules_string) 
+        + az_span_size(client->_internal.options.module_id);
+  }
+
+  if (az_span_size(key_name) > 0)
+  {
+    required_size += 
+        + (int32_t)sizeof(AMPERSAND)
```

Comment:
> +  [](start = 7, length = 3)

+?

Reply:
yeah, I think I saw other comments before of people in our team preferring the explicit if block. I know it's bigger, but more people prefer it. Sorry.